### PR TITLE
the Safe Update

### LINKED ISSUE #25720: Safe bug is still here and it's mean and lean
(WEB REPORT BY: casualbro REMOTE: 206.221.180.138:7777)
Something about how the tumbler positions are dragged along seems to be responsible for the rare chance that the safe breaks. A gap between the opening positions before the second tumbler unlocks and starts decrementing is responsible, but changing the culprit values from 35 to 37 broke safe-cracking, so I'm leaving this here for either someone else to fix or for me to fix

### LINKED ISSUE #28027: Safes can spawn impossible to open due to overlapping tumbler positions
(WEB REPORT BY: loppu REMOTE: 206.221.180.138:7777)
# Revision
a6892f543e5ef978de5818be03c4bed26b0d72c9

# Description
Safes can spawn with their target tumbler positions too close together making them impossible to open due to overlap. The position to get the tumblers moving again after locking one will be after the target position to lock the other tumbler, making it impossible to reach without resetting the tumblers.

# Steps to Reproduce
Get an unlucky safe spawn.

# What you Expected
This is a bug I recognize from years ago that was fixed until a few days ago.

# What Actually Happened
Unsurprisingly when I look up the safe code on github shiftys clothing datums are listed as the most recent changes. Thanks obama.

### PR DESCRIPTION
**Pull request by DeityLink**
image
Gotta preserve the traditions.

Anyway, Safe UI. Safe SFX. Safe revamp.

Fixes #25720
Fixes #28027

safeUI

https://user-images.githubusercontent.com/7573912/119910722-06f9e600-bf58-11eb-8c40-b99ae358b967.mp4

:cl:
* rscadd: The Safe has had its functionality and UI refreshed a bit. Instead of bringing up an annoying html pop-up, clicking the Safe now brings up a radial menu featuring the safe's dial at the center and two buttons to rotate it. Immersion! Furthermore the Safe now opens by itself as soon as both tumblers are properly aligned and its contents will immediately appear on top of it. Clicking the safe while it's open with an item will place said item on top of it, while clicking it with an empty hand will close it, moving the items on top inside it, and randomize the tumblers' positions.
* rscadd: You can now use a screwdriver on an open Safe to randomize the tumblers' unlock positions if you want.
* tweak: The Safe's maximum capacity is now 30 items of any w_class, just like crates and closets, instead of its max capacity being limited by the items' combined w_class.
* soundadd: The Safe now makes some sounds! A few brand new SFXs have been added.
* bugfix: Fixed a very rare bug where the Safe's tumblers were set in a way that prevented players from ever cracking the Safe open.

### CODE REVIEW

**Comment by beesil** [comment]
get your closevalids on #25720 and #28027!

**Comment by help-maint** [comment]
sixth

**Comment by DeityLink** [comment]
@help-maint I observed it happening like two days ago to Eneocho, one of the tumbler's unlock position was 70. The max is now 69 just to be safe.

@boy2mantwicethefam your fix technically halved the occurrence of the bug because you upped the min value, but you forgot to lower the max value.

**Comment by Eneocho** [comment]
> 
> 
> @help-maint I observed it happening like two days ago to Eneocho, one of the tumbler's unlock position was 70. The max is now 69 just to be safe.

That safe was fucking frustrating. The moment I managed to set the first tumbler, the second tumbler unset itself.

**Comment by Eneocho** [comment]
Does this mean you can move safes? If not, please make it an option similar to moving an anvil.
Also, do the klink, tonk and whatever on the chat still happen? Or are they completely replaced by the sound system? 
Is there an SFX for setting the first or second tumbler?

File: code/_onclick/hud/radial.dm [hunk 1]
```diff
@@ -313,31 +331,76 @@
 		qdel(custom_check)
 		custom_check = null
 	. = ..()
+
+/////////////////////////////////////////Recursive radial menu, currently used by Safes
+
+/datum/radial_menu/proc/do_it_again()
+	selected_choice = null
+	if (istype(anchor,/obj/structure/safe))
```
Comment (help-maint) [inline code/_onclick/hud/radial.dm]: This functionality should probably be moved into a radial_menu/safe subtype, rather than typechecking here (muh OOP)

**Comment by Kurfursten** [comment]
I'm waiting for Safe 3.0 with real security pin action.

"I'm going to be using this wiper insert with bottom of the keyway tension. Then, I'll be using this three thousands pick from the Covert Companion. Nothing on one. Little click out of two. Three is binding. And we've fallen into a false set, just doubling back... and open."

**Comment by Eneocho** [comment]
> 
> 
> I'm waiting for Safe 3.0 with real security pin action.
> 
> "I'm going to be using this wiper insert with bottom of the keyway tension. Then, I'll be using this three thousands pick from the Covert Companion. Nothing on one. Little click out of two. Three is binding. And we've fallen into a false set, just doubling back... and open."

Seems like the logical conclusion to all the LPL videos. Him cracking a space vault safe.

**Comment by DeityLink** [comment]
Alright the latestet improvements should prevent clogging of the radial menu altogether but I'm keeping the procs in just in case. Better safe than sorry. Should be good for review.

File: code/_onclick/hud/radial.dm (hunk 1 above)
Comment (DeityLink) [reply code/_onclick/hud/radial.dm]: Yep, the type check is now done only once when the menu is created, and the dial subtype will update the icon.

**Comment by DeityLink** [comment]
@Eneocho 

> Does this mean you can move safes? If not, please make it an option similar to moving an anvil.

Maybe in a later PR, this one already swallowed an entire day for me.

> Also, do the klink, tonk and whatever on the chat still happen? Or are they completely replaced by the sound system?

They still happen. Ideally in a later PR they should pop above the safe goonchat-style. Maybe color-coded so you know if they come from the 1st or 2nd tumbler too.

> Is there an SFX for setting the first or second tumbler?

There are 3 new SFXs. One is a simple click always heard when you move the dial. One is a slightly more echo-y click heard when at least one tumbler is aligned with their unlock position (but that can only be heard when holding a stethoscope still). The third one is a fast series of clicks heard when closing the safe and resetting the tumblers.

**Comment by help-maint** [comment]
Wait, without the popup window, how do I tell what number the dial is set to? Say, if I want to tell someone else that the code is 60-210

**Comment by Eneocho** [comment]
> 
> 
> Wait, without the popup window, how do I tell what number the dial is set to?

There is no number, there's the dial angle you can visually see. It could have a digital 7 segment square above the wheel that tells you the current position number though.

**Comment by DeityLink** [comment]
> 
> 
> > Wait, without the popup window, how do I tell what number the dial is set to?
> 
> There is no number, there's the dial angle you can visually see. It could have a digital 7 segment square above the wheel that tells you the current position number though.

how 'bout this 
image

**Comment by Eneocho** [comment]
> 
> how 'bout this
> image

I would have liked the seven segment display as an immulshon in-screen thing, but this works as well.

**Comment by DeityLink** [comment]
Also by the way the tumblers now only make their unlocked sound (like "tink", "krink" or "plink" for the 2nd tumbler) if they just turned. So no more spam of those when changing direction and not actually moving them. This might make you think that the other tumbler went out of position but you'll open the safe just fine once both tumblers are aligned.

image
*later*
image

**Comment by Eneocho** [comment]
It rustles me jimmies that the dial overlay isn't centered above the Safe's dial.

**Comment by help-maint** [comment]
Just tossing around ideas, but you could maybe display the dial number by setting a maptext on the radial's screen element?

'sides that, looks good

**Comment by DeityLink** [comment]
> 
> 
> It rustles me jimmies that the dial overlay isn't centered above the Safe's dial.

The floor Safe has its dial centered differently as well, you'll get over it.


> Just tossing around ideas, but you could maybe display the dial number by setting a maptext on the radial's screen element?

I've already considered that idea, I did not like it aesthetically.

**Pull request merged**